Update mesh expansion doc with birectional example. (#3274)
istio/istio.io: content/docs/setup/kubernetes/additional-setup/mesh-expansion/index.md

include exposing the Kubernetes DNS server through an internal load balancer, using a Core DNS
server, or configuring the IPs in any other DNS server accessible from the VM.

* Mesh expansion has to be enabled by Helm. You need to install the [Helm client](https://docs.helm.sh/using_helm/).

> {{< tip >}}

Author: Removed

> * If you haven't already enabled mesh expansion [at install time with Helm](/docs/setup/kubernetes/install/helm/), you have

Author: Done.

generate the configuration files that let mesh expansion VMs connect to the mesh. To prepare the
cluster for mesh expansion, run the following commands on a machine with cluster admin privileges:

1.  Ensure that mesh expansion is enabled for the cluster. If you did not specify `--set global.meshExpansion.enabled=true` at
    install time with Helm, there are two options for enabling mesh expansion, depending on how you originally installed
    Istio on the cluster:

    *   If you installed Istio with Helm and Tiller, run `helm upgrade` with the new option:

    {{< text bash >}}
    $ cd install/kubernetes/helm/istio
    $ helm upgrade --set global.meshExpansion.enabled=true istio-system .
    $ cd -
    {{< /text >}}

>     *   If you installed Istio without Helm and Tiller, use `helm template` to update your configuration with the option and reapply with `kubectl`:

Author: Yes. I think that's accurate given the mesh expansion option is only available as a helm value flag.

>     *   If you installed Istio without Helm and Tiller, use `helm template` to update your configuration with the option and reapply with `kubectl`:

Author: I updated the prerequisite sections to reflect `helm` client is a must anyway.

          -o jsonpath='{.data.cert-chain\.pem}' |base64 --decode > cert-chain.pem
    {{< /text >}}

### Setting up the VM

Next, run the following commands on each machine that you want to add to the mesh:

> <!-- TODO: find a better way to reference release in the URL instead of manually sync... -->

Author: Filed https://github.com/istio/istio.io/issues/3496

    $ sudo systemctl start istio
    {{< /text >}}

> ## VM workloads send request to Kubernetes services

Reviewer: ```suggestion
## Send requests from VM workloads to Kubernetes services
```

1.  First, on the cluster admin machine get the virtual IP address (`clusterIP`) for the service:

    {{< text bash >}}
    $ kubectl get svc productpage -o jsonpath='{.spec.clusterIP}'
    10.55.246.247
    {{< /text >}}

1.  Then on the mesh expansion machine, add the service name and address to its `etc/hosts` file. You can then connect to
    the cluster service from the VM, as in the example below:

    {{< text bash >}}
$ echo "10.55.246.247 productpage.default.svc.cluster.local" | sudo tee -a /etc/hosts
$ curl -v productpage.default.svc.cluster.local:9080
< HTTP/1.1 200 OK
< content-type: text/html; charset=utf-8
< content-length: 1836
< server: envoy
... html content ...
    {{< /text >}}

> The `server: envoy` header indicates the traffic indeed is intercepted by sidecar.

Reviewer: ```suggestion
The `server: envoy` header indicates that the sidecar intercepted the traffic.
```

## Running services on a mesh expansion machine

> 1. We first setup a HTTP server on the VM instance, serving HTTP traffic on port 8080.

Reviewer: ```suggestion
1. Setup an HTTP server on the VM instance to serve HTTP traffic on port 8080:
```

    {{< text bash >}}
    $ gcloud compute ssh ${GCE_NAME}
    $ python -m SimpleHTTPServer 8080
    {{< /text >}}

> 1. Determine the VM instance IP address. For example, commands to find GCE instance IP are as below.

Reviewer: ```suggestion
1. Determine the VM instance's IP address. For example, find the IP address
    of the GCE instance with the following commands:
```

    {{< text bash >}}
    $ export GCE_IP=$(gcloud --format="value(networkInterfaces[0].networkIP)" compute instances describe  ${GCE_NAME})
    $ echo ${GCE_IP}
    {{< /text >}}

> 1. Configure Service Discovery for VM by `ServiceEntry`. You add VM services to the mesh by configuring a

Reviewer: Please update the entire text of the step to avoid clunky content for our users.
```suggestion
1. Configure a service entry to enable service discovery for the VM. You can add VM services to the mesh using a 
    [service entry](/docs/reference/config/networking/v1alpha3/service-entry/). Service entries let you manually add
    additional services to Pilot's abstract model of the mesh. Once VM services are part of the mesh's abstract model,  
    other services can find and direct traffic to them. Each service entry configuration contains the IP addresses, ports, 
    and appropriate labelsof all VMs exposing a particular service, for example:
```

Author: Adopted. but do we want to break the word `ServiceEntry`, which is a specific terminology in Istio?

Reviewer: Yes, we do. See:  https://preliminary.istio.io/about/contribute/style-guide/#use-consistent-capitalization for details.

[`ServiceEntry`](/docs/reference/config/networking/v1alpha3/service-entry/). A `ServiceEntry` lets you manually add
additional services to Istio's model of the mesh so that other services can find and direct traffic to them. Each
`ServiceEntry` configuration contains the IP addresses, ports, and labels (where appropriate) of all VMs exposing a
particular service, as in the following example.

    {{< text bash yaml >}}
    $ kubectl -n ${SERVICE_NAMESPACE} apply -f - <<EOF
    apiVersion: networking.istio.io/v1alpha3
    kind: ServiceEntry
    metadata:
    name: vmhttp
    spec:
    hosts:
    - vmhttp.${SERVICE_NAMESPACE}.svc.cluster.local
    ports:
    - number: 8080
        name: http
        protocol: HTTP
    resolution: STATIC
    endpoints:
        - address: ${GCE_IP}
        ports:
            http: 8080
        labels:
            app: vmhttp
            version: "v1"
    EOF
    {{< /text >}}

> 1. Workloads in a Kubernetes cluster need a DNS mapping to resolve the service running on the VM.

Reviewer: ```suggestion
1. The workloads in a Kubernetes cluster need a DNS mapping to resolve the the domain names of VM services. To 
    integrate the mapping with you own DNS system, use `istioctl register` and creates a Kubernetes `selector-less` 
    service, for example:
```

Author: Although not matter in markdown, why newline in these cases? I searched a bit on [style guide](https://preliminary.istio.io/about/contribute/style-guide) but didn't find any recommended line length.

Reviewer: This is a recommendation from GitHub. Up to 80 characters are properly shown in a single line for review.

You can integrate with your own DNS system. For illustration purpose, we use `istioctl register`
which creates a Kubernetes `selector-less` service.

    {{< text bash >}}
    $ istioctl  register -n ${SERVICE_NAMESPACE} vmhttp ${GCE_IP} 8080
    {{< /text >}}

> 1. We deploy a sleep pod in the Kubernetes cluster, and wait for the pod to be ready.

Reviewer: ```suggestion
1. Deploy a pod running the `sleep` service in the Kubernetes cluster, and wait until it is ready:
```

    {{< text bash >}}
    $ kubectl apply -f samples/sleep/sleep.yaml
    $ kubectl get po
    NAME                             READY     STATUS    RESTARTS   AGE
    productpage-v1-8fcdcb496-xgkwg   2/2       Running   0          1d
    sleep-88ddbcfdd-rm42k            2/2       Running   0          1s
    ...
    {{< /text >}}

> 1. Send a request from the sleep container to the VM HTTP service.

Reviewer: ```suggestion
1. Send a request from the `sleep` service on the pod to the VM's HTTP service:
```

    {{< text bash >}}
    $ kubectl exec -it sleep-88ddbcfdd-rm42k -c sleep -- curl vmhttp.${SERVICE_NAMESPACE}.svc.cluster.local:8080
    <!DOCTYPE html PUBLIC "-//W3C//DTD HTML 3.2 Final//EN"><html>
    <title>Directory listing for /</title>
    <body>
    <h2>Directory listing for /</h2>
    <hr>   
    <ul>                                                                
    <li><a href=".bashrc">.bashrc</a>
    <li><a href=".ssh/">.ssh/</a>
    </body>
    {{< /text >}}

> 

Reviewer: Add text letting the readers know what they achieved:
```suggestion

**Congratulations!**

You successfully configured a service running in a pod within the mesh to send 
traffic to an external service running on a VM outside of the mesh and 
tested that the configuration worked. 

```

Author: Thanks for pointing out, this is helpful :)

But the wording is not quite  accurate right? The whole point of the `mesh expansion` is to make VM part of the mesh beyond Kubernetes clusters. So the VM are also part of the mesh, not external of it. Or your intention is to replace that with "Kubernetes" cluster?

Reviewer: I agree, you are adding the VM to the mesh but not the cluster. Changing "mesh" to cluster on the second line of my suggestion should clarify.

Author: Gotcha. Done.

## Cleanup

> Run the following commands to remove the expansion VM from the mesh.

Reviewer: ```suggestion
Run the following commands to remove the VM services from the mesh's abstract model.
```

Author: Given the page is titled as "mesh expansion" I think "remove the VM from the mesh" is clear enough? Any reason to mention "abstract model"? I'm not sure if we have settled down a model for VM.

Reviewer: So, the abstract model is not the model of the VM but the abstract model Pilot maintains for the mesh. Remove VM from mesh is not false but it's too general. The added specificity helps readers understand the behavior.

Author: Done.

... html content ...
    {{< /text >}}

The `server: envoy` header indicates that the sidecar intercepted the traffic.

## Running services on a mesh expansion machine

1. Setup an HTTP server on the VM instance to serve HTTP traffic on port 8080:

    {{< text bash >}}
    $ gcloud compute ssh ${GCE_NAME}
    $ python -m SimpleHTTPServer 8080
    {{< /text >}}

1. Determine the VM instance's IP address. For example, find the IP address
    of the GCE instance with the following commands:

    {{< text bash >}}
    $ export GCE_IP=$(gcloud --format="value(networkInterfaces[0].networkIP)" compute instances describe  ${GCE_NAME})
    $ echo ${GCE_IP}
    {{< /text >}}

1. Configure a service entry to enable service discovery for the VM. You can add VM services to the mesh using a
    [service entry](/docs/reference/config/networking/v1alpha3/service-entry/). Service entries let you manually add
    additional services to Pilot's abstract model of the mesh. Once VM services are part of the mesh's abstract model,
    other services can find and direct traffic to them. Each service entry configuration contains the IP addresses, ports,

>     and appropriate labels of all VMs exposing a particular service, for example:

Reviewer: Thanks for accepting my suggestion but please remove lines 237-238.

    EOF
    {{< /text >}}

> 1. The workloads in a Kubernetes cluster need a DNS mapping to resolve the the domain names of VM services. To

Reviewer: ```suggestion
1. The workloads in a Kubernetes cluster need a DNS mapping to resolve the domain names of VM services. To
```

Author: Done.

    integrate the mapping with you own DNS system, use `istioctl register` and creates a Kubernetes `selector-less`

>     service, for example:

Reviewer: Thanks for accepting my suggestion but please remove lines 269-270.

Author: Done.

    <body>
    <h2>Directory listing for /</h2>
    <hr>
    <ul>
    <li><a href=".bashrc">.bashrc</a>
    <li><a href=".ssh/">.ssh/</a>
    </body>
    {{< /text >}}

**Congratulations!**

You successfully configured a service running in a pod within the mesh to send
traffic to an external service running on a VM outside of the mesh and
tested that the configuration worked.

> **Congratulations!**

Reviewer: Thanks for accepting my suggestion but please remove lines 309-311.

Author: Done.
